### ISSUE TITLE
0.10.0 upgrade guide

### ISSUE THREAD

**Issue by davidchambers** [ramda/ramda#812]
Changes: https://github.com/ramda/ramda/compare/v0.9.1...v0.10.0

Name changes:

  - `R.mixin` → `R.merge` #783
  - `R.foldl` → `R.reduce` #769 #775
  - `R.foldr` → `R.reduceRight` #769 #775
  - `R.foldlIndexed` → `R.reduceIndexed` #775
  - `R.foldrIndexed` → `R.reduceRightIndexed` #775
  - `R.lPartial` → `R.partial` #775
  - `R.rPartial` → `R.partialRight` #775
  - `R.mapAccumL` → `R.mapAccum` #775
  - `R.mapAccumR` → `R.mapAccumRight` #775
  - `R.scanl` → `R.scan` #775
  - `R.unfoldr` → `R.unfold` #775
  - `R.pCompose` → `R.composeP` #775
  - `R.pPipe` → `R.pipeP` #775

Improvements:

  - `R.construct` and `R.constructN` now support the Date constructor #761
  - `R.invert` now ignores inherited properties #795
  - `R.lift` now works with Fantasy Land Applicatives #782
  - `R.memoize` now reuses cached "false" values #785
  - `R.memoize` once again works with nullary functions #787
  - :curry: `R.merge` (formerly `R.mixin`) is now curried op-style #745
  - :warning: `R.of` no longer takes an optional second argument #784
  - `R.reject` now dispatches to "filter" method if present #799
  - `R.sortBy` now preserves object identity #789

Additions:

  - :new: `R.defaultTo` #788
  - :new: `R.dissoc` #781
  - :new: `R.mergeAll` #778

**Comment by morlay** [comment]
Thanks for the awesome work. I like ramda and use it in some product now.
But please don't remove apis directly like this. Maybe use some warning will be better, and keep them as  alias.
( That's true, we just need to rename for upgrading. howerever, not user-friendly enough. is it? :-)

**Comment by davezuko** [comment]
We've been trying to incorporate Ramda into our codebase but with the latest updates it's getting pretty scary (reduce randomly removed, mixin renamed to merge). As Morlay said it would be nice, until a major version is released, to at least maintain some aliases so we can spend more time using Ramda and less time fixing broken methods.

**Comment by Aourin** [comment]
I get that it still isn't technically a full release, but the changes get kind of crazy with the loss of the api without alias. I agree with @morlay . Could you guys/gals at least alias while providing a deprecation flag until the next release. That way people know to start changing their stuff. It is starting to make me not want to use Ramda until it is in at a full version release.

**Comment by TheLudd** [comment]
I ask you to reconsider #786

**Comment by CrossEye** [comment]
I think we may well have to reconsider something like that, @TheLudd.

But I also think we're almost at the end of this.  We're hoping to hit a 1.0 release in a few months, and after that we expect to be following [semver][sv] scrupulously.  At the moment, we're trying to make sure our API is actually in shape to be released as a 1.0 version.  Some of that involves the kind of normalization that's gone on in the last two releases.  We've chosen to do this without the safety net of aliases or deprecation warnings, because, as @davidchambers points out, it should generally be easy to incorporate such changes with a global search-and-replace.

I know it's not ideal.  And if we have another major round of it, we'll reconsider the approach.  But I do think we're almost done with it.

Sorry.


  [sv]: http://semver.org/

**Comment by YBNSalesEngineering** [comment]
I agree with @Aourin. @CrossEye - I understand that you are working towards a 1.0 version but you should definitely follow semver now as many of your users are incorporating this library into their projects. Please consider keeping aliases until you reach a major version.

**Comment by davidchambers** [comment]
> you should definitely follow semver

We do. From the [spec][1]:

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.


[1]: http://semver.org/spec/v2.0.0.html

**Comment by Luckvery** [comment]
@davidchambers Before I say anything, let me take a moment to thank you for all you do to make ramda an awesome library. However...

If you are going to quote the spec, please do not do so out of context:

> How do I know when to release 1.0.0?

> If your software is being used in production, it should probably already be 1.0.0. If you have a 
> stable API on which users have come to depend, you should be 1.0.0. If you're worrying a lot 
> about backwards compatibility, you should probably already be 1.0.0.

TL:DR Don't diss us who love ramda enough to be using it in production already. Aliasing is a simple solution.

**Comment by buzzdecafe** [comment]
I'd say we meet1/3 of these requirements:

* your software is being used in production :heavy_check_mark:  
* you have a stable API on which users have come to depend :x:
* you're worrying a lot about backwards compatibility :x:

My apologies for the growing pains, but we really need the pre-1.0 time to tinker with things and get them as right as we can.

**Comment by CrossEye** [comment]
@buzzdecafe:

I think that list was supposed to be: "If any of these conditions is true..."

I think you and I are having a harder time because although we are the initial authors of the library, we are not using it in our day job yet.  I suspect @davidchambers is doing so but that his process lets him very easily migrate versions.

This is making me rethink temporary aliases as at least a short-term solution, either on our own or by promoting [Aourin's Ramda-Aliases][ga].

  [ga]: https://github.com/Aourin/ramda-alias

**Comment by davidchambers** [comment]
I'm thrilled to hear, @Luckvery, that I'm not the only one using Ramda in production. :)

I've been using Ramda in production since `0.2.x`. We now use it in more than a dozen projects at [Plaid][1]. In the past 48 hours I opened 29 pull requests upgrading Ramda in these projects (some were still on `0.7.x` or `0.8.x` so required multiple upgrades). These upgrades took ten minutes each, on average.

I find a ten-minute upgrade acceptable for Ramda at this point in its life. Since upgrading is voluntary, we're free to continue to use whichever version we're currently using until we can devote some time to upgrading. When we decide to upgrade, the task should not be onerous.

Upgrading from `0.n.x` to `0.(n + 1).x` should never be done blindly. Even in a release in which no functions are renamed there could be any number of breaking changes. The upgrade process is thus very similar regardless of whether a release includes name changes:

1.  Preparation:

    ```console
    $ cd path/to/project
    $ git checkout master
    $ git pull origin master
    $ git checkout -b ramda
    $ vim package.json  # update Ramda dependency
    $ rm -r node_modules
    $ npm install  # install the new version
    ```

2.  Read the [upgrade guide][2] one line at a time, making any necessary changes. `ack` and `sed -i ''` are very useful here.

3.  Run tests, if present:

    ```console
    $ npm test
    ```

4.  Address any issues discovered in the previous step.

5.  Commit and push:

    ```console
    $ git add .
    $ git commit -m 'ramda@0.10.x'
    $ git push origin ramda
    ```

6.  Open a pull request. Include a link to the [upgrade guide][2] in the description to provide context for reviewers.

What would including aliases for renamed functions buy us? It would reduce the number of changes required in step 2.

Renaming functions is the easiest part of this step:

```console
$ find lib -name '*.js' -print0 | xargs -0 sed -i '' 's/R[.]mixin/R.merge/g'
```

Replace `mixin` and `merge` with the next old–new pair and repeat.

Other sorts of change require more attention. For example, the order of arguments to `R.propOr` changed in 0.9.0. Easily the scariest changes are semantic changes. When `R.isEmpty` was updated to return false when applied to null/undefined, we had to review every call site to ensure the function is only ever applied to an array-like object, and add `xs == null ||` in sites which required the old behaviour.

To me, having to rename functions with each minor release until we reach 1.0.0 is not a big cost in absolute terms, nor is it a big cost relative to the other code changes upgrades sometimes require.

Does one of you have an open-source project using an old version of Ramda? I'm interested in understanding what it is that's making upgrading a daunting task in some cases. If you provide a link to the project I'll submit a pull request and we can discuss the pain points.


[1]: https://plaid.com/
[2]: https://github.com/ramda/ramda/issues/812

**Comment by buzzdecafe** [comment]
@davidchambers that answer might be useful boilerplate for future upgrade guides

**Comment by Luckvery** [comment]
@davidchambers 
I think you make a valid point about upgrading when you feel like it. Also, changes to our projects flow very similar to yours. 

> In the past 48 hours I opened 29 pull requests upgrading Ramda in these projects (some were still on0.7.x or 0.8.x 
> so required multiple upgrades). These upgrades took ten minutes each, on average.

So if my math serves me that is 29 X 10 minutes totaling 290 minutes or just under 5 hours. If that works for you than I don’t think I am skillful enough to construct an argument that would sway your opinion anyway lol.  

>  Easily the scariest changes are semantic changes. 

I couldn’t agree with you more. Aliasing just means that I would only have to focus on these semantical changes. We currently have ramda in production in 2 small projects (including tests) and are now using it in our largest project for our most demanding (read we want changes yesterday type) client.

> To me, having to rename functions with each minor release until we reach 1.0.0 is not a big cost in absolute terms, nor 
> is it a big cost relative to the other code changes upgrades sometimes require.

 When managing several projects  in (what some would consider) a fast paced environment the frequent changes to ramda aren’t _daunting_ they are just a pita i’d like to avoid if I could. 

oh well.  I'm still going to use ramda anyway.

**Comment by davidchambers** [comment]
Thank you for elaborating your thoughts, @Luckvery.

> I think you make a valid point about upgrading when you feel like it.

The main problem with this defence (if I may undermine my own defence) is that we don't yet expose documentation for old versions (see #821). A mismatch between API and documentation is no fun.

> When managing several projects in (what some would consider) a fast paced environment the frequent changes to ramda aren’t *daunting* they are just a pita i’d like to avoid if I could.

Fair point. There's a trade-off between ease of upgrading and simplicity of Ramda's internals. [Aourin/ramda-alias][1] keeps the aliases out of Ramda and makes upgrading easier. It seems like a good solution until we hit 1.0.0.


[1]: https://github.com/Aourin/ramda-alias/pull/1/files

**Comment by jdalton** [comment]
It's the early adopter cost of working with a pre 1.0 lib. Ramda has been pre 1.0 since 2013 so users should be used to this kind of shift and change. Libs need the freedom to adjust in their pre 1.0 period to iron things out and create a better experience for their users without having the back-compat burden.

**Comment by davezuko** [comment]
There's something to be said about sitting behind the pre-1.0 safety net for multiple years though. I'm not complaining, and we'll still continue to use Ramda in our production environments, but imagine if Node took a similar approach and started introducing massive breaking changes just because they're pre-1.0. Early adopters of Ramda will make do because they want to use Ramda, so they'll make the appropriate adjustments to the API updates. The argument is more about general reliability and ease-of-upgrade than anything.

**Comment by jdalton** [comment]
You can't really compare Node as it has it's own versioning scheme and doesn't follow semver. Early adopters are generally super fans which follow the project's changes and activity closely and as such should cut the maintainers some slack. As an early adopter of a lib that's been unstable for ~2 years you've chosen style over stability and part of that means dealing with breaking changes. Ramda could as a courtesy bump the minor version when introducing breaking changes so the `~` version range would work (they may already be doing that).

**Comment by davidchambers** [comment]
> Ramda could as a courtesy bump the minor version when introducing breaking changes so the `~` version range would work (they may already be doing that).

That's our approach, yes. I use `0.9.x`, `0.10.x`, etc. when specifying my Ramda dependencies.

**Comment by CrossEye** [comment]
The plan is to hit 1.0 fairly soon.

Ramda started as a personal project, a way to learn this stuff, by @buzzdecafe and me, with an [earlier library][ew] back in April 2013.  By that June, we'd learned a lot and realized that if we were going to have a practical library, we were going to have to sacrifice much of the FP elegance.  So we created Ramda in June, 2013.  But it was still just our own playground for quite some time.  There were a few GitHub watchers, mostly our friends, but no real activity except that piqued by our own interests.

Then last May, @buzzdecafe and I were in Frontend Masters' course, [Hardcore Functional Programming in JavaScript][hc] only to find that they were using our library for the main examples!

That gave us the courage to [announce][an] Ramda publicly last May.  And it's been a wild ride ever since.

Sometime around either the second anniversary of the birth of Ramda (in June) or the first anniversary of its announcement (May), I'm hoping that we can be ready to release a 1.0 version.  We seem to be getting pretty close right now, due to the hard work of all our contributors, but most especially to @davidchambers' eye for detail.

We seem to be converging on a very useful and clean API, our performance is in line with other similar libraries, and there is clearly a desire for this sort of tool.

So I don't think we'll be in this sort of limbo much longer.


  [an]: http://buzzdecafe.github.io/code/2014/05/16/introducing-ramda/
  [ew]: https://github.com/CrossEye/eweda
  [hc]: https://frontendmasters.com/courses/functional-javascript/

**Comment by kedashoe** [comment]
Are you guys not adding the release notes to the release page anymore? Is there another easier-than-searching-the-issues place to find them?

**Comment by elijahdorman** [comment]
Is there any particular reason for not to add the new names and simply deprecate the old ones It doesn't add that much to the size of the library, but keeps a lot of breakages from happening until we all catch up.

**Comment by CrossEye** [comment]
@kedashoe (sorry, too much activity, and missed this a few days ago): If we haven't been keeping up the release page, it's been an oversight.  I think we're going to try to keep that up as well as a changelog.

@elijahdorman: The [parable][pb] presented a while ago convinced most of us that aliases were to be used very sparingly, if at all.  We have seemed to agree to run a version or two with a deprecation warning for a function that's due to be replaced.  (We haven't really talked about whether we'd use an alias during this transition.)  But we're really trying to reach the point where none of this makes much of a difference, as we've stopped significant API churn.  


  [pb]: https://github.com/ramda/ramda/pull/574#issuecomment-64124850
